### PR TITLE
Refactoring and support PaliGemma 

### PR DESCRIPTION
**Pull request by mshukor**
## What this does
This PR inlcudes:
* Organize the branch so that we can easily switch action heads and vlms
* Support PaliGemma 3B models
* Fix some issues related to prompt templates

## How it was tested
```

SCRATCH=/tmp/log_dir

ENV=aloha
ENV_TASK=AlohaTransferCube-v0
dataset_repo_id=lerobot/aloha_sim_transfer_cube_human



EVAL_FREQ=5000 
OFFLINE_STEPS=100000 
TRAIN_BATCH_SIZE=8
EVAL_BATCH_SIZE=8
SAVE_FREQ=20000

policy=vla
LR=1e-5
USE_AMP=true
ASYNC_ENV=false

## Select different VLMs features selection method
# FEAT_SELECT=last_token
FEAT_SELECT=all_generated


## Select different VLMs
# VLM=llava-hf/llava-onevision-qwen2-0.5b-ov-hf
# VLM_NAME=llavaonevision_05b
# VLM_DIM=896

VLM=google/paligemma2-3b-pt-224
VLM_NAME=paligemma2_3b
VLM_DIM=2304


TASK_NAME=lerobot_${ENV}_transfer_cube_${policy}_${VLM_NAME}_feat_select_${FEAT_SELECT}


MUJOCO_GL=egl python lerobot/scripts/train.py \
 hydra.job.name=base_distributed_aloha_transfer_cube \
 hydra.run.dir=$SCRATCH/logs/lerobot/${TASK_NAME} \
 dataset_repo_id=$dataset_repo_id \
 policy=$policy \
 env=$ENV env.task=$ENV_TASK \
 training.offline_steps=$OFFLINE_STEPS training.batch_size=$TRAIN_BATCH_SIZE training.save_freq=$SAVE_FREQ \
 training.eval_freq=$EVAL_FREQ eval.n_episodes=50 eval.use_async_envs=$ASYNC_ENV eval.batch_size=$EVAL_BATCH_SIZE \
 training.lr=$LR \
 wandb.enable=true use_amp=$USE_AMP \
 policy.vlm_backbone.feature_selection=$FEAT_SELECT policy.vlm_backbone.name=$VLM policy.action_decoder.dim_model=$VLM_DIM

```

### CODE REVIEW

File: lerobot/common/policies/vla/modeling_vla.py [hunk 1]
```diff
@@ -78,299 +66,210 @@ def reset(self):
...
 
-    def forward(self, processed_inputs, batch_size):
+        if any(k in self.vlm_backbone_name for k in ["llava-onevision", "paligemma"]):
+            batch_size = processed_inputs["input_ids"].shape[0]
+            last_hidden_state = vlm_output.hidden_states[-1]
+            seq_len = vlm_output.image_hidden_states.shape[0] // batch_size
+            image_features = vlm_output.image_hidden_states.view(batch_size, seq_len, -1)
+
+            if self.vlm_backbone_feature_selection == "first_image":
+                num_img_feats = 598  # this is specific to llava-onevision
```
Comment (danaaubakirova) [inline lerobot/common/policies/vla/modeling_vla.py:207]: we can move this to config as well. not really specific to llava. As I mentioned this was to match the number of hidden states in ACT

File: lerobot/common/policies/vla/modeling_vla.py [hunk 2]
```diff
@@ -78,299 +66,210 @@ def reset(self):
...
+                hidden_states = image_features[:, :num_img_feats, :]
+            elif self.vlm_backbone_feature_selection == "last_token":
+                hidden_states = last_hidden_state[:, -1:, :]
+            elif self.vlm_backbone_feature_selection == "all_generated":
+                hidden_states = last_hidden_state
+            elif self.vlm_backbone_feature_selection == "all":
+                hidden_states = torch.cat((image_features, last_hidden_state), dim=1)
```
Comment (danaaubakirova) [inline lerobot/common/policies/vla/modeling_vla.py:214]: I think we should have some hyperparameter to control the number of image features, and that would eliminate the need for the first if statement? wdyt

File: lerobot/common/policies/vla/modeling_vla.py [hunk 3]
```diff
@@ -78,299 +66,210 @@ def reset(self):
...
+                (self.chunk_size, batch_size, hidden_size),
+                dtype=hidden_states.dtype,
+                device=hidden_states.device,
+            )
+            action_logits = self.action_decoder(
+                x=x, encoder_out=hidden_states, decoder_pos_embed=decoder_pos_embeddings
+            )
+            # Final action logits through the action head
+            action_logits = self.action_head(action_logits)
+            action_logits = action_logits.transpose(0, 1)
```
Comment (danaaubakirova) [inline lerobot/common/policies/vla/modeling_vla.py:239]: that's looks good to me:) Indeed would be better if we want to add other decoders.

**Review by danaaubakirova — COMMENTED** [review]
Thanks for this PR! Refactoring LGTM! Thanks for deleting the unnecessary files. I want to experiment with faster processors (new integrated feature in transformers). That might change the code a bit. But not significantly.

File: lerobot/common/policies/vla/modeling_vla.py [hunk 4]
```diff
@@ -78,299 +66,210 @@ def reset(self):
...
         # Verify trainable parameters
         trainable_params = []
         for name, param in self.vision_language_model.named_parameters():
             if param.requires_grad:
                 trainable_params.append(name)
                 print(f"Trainable parameter: {name}")
 
-        #self.device = self.vision_language_model.device
-        #self.half()
+    def apply_prompt_template(self, text: str, add_generation_prompt: bool = True) -> str:
```
[the remark this answers is not in the document]
Comment (mshukor) [reply lerobot/common/policies/vla/modeling_vla.py:174]: Some VLMs are trained on SFT data with specific template. This template should be used to get the best answers. I will add parameter to test if it makes any difference here

File: lerobot/common/policies/vla/modeling_vla.py [hunk 5]
```diff
@@ -78,299 +66,210 @@ def reset(self):
...
+        elif "paligemma" in self.vlm_backbone_name:
+            from transformers import PaliGemmaForConditionalGeneration, PaliGemmaProcessor
+
+            self.vision_language_model = PaliGemmaForConditionalGeneration.from_pretrained(
+                self.vlm_backbone_name,
+                device_map="auto",
+                torch_dtype=torch.float16,
+                # attn_implementation="flash_attention_2"
+            )
+            self.processor = PaliGemmaProcessor.from_pretrained(self.vlm_backbone_name)
```
[the remark this answers is not in the document]
Comment (mshukor) [reply lerobot/common/policies/vla/modeling_vla.py:143]: yes

File: lerobot/common/policies/vla/modeling_vla.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mshukor) [reply lerobot/common/policies/vla/modeling_vla.py:177]: But you should also explicitly add <image> right? it was not there and was throwing an error for me

[the remark this answers is not in the document]
Comment (danaaubakirova) [reply lerobot/common/policies/vla/modeling_vla.py:177]: I see now. I was using the older version of transformers before they integrated chat template. Yes, then we need to add it explicitly.

**Pull request merged**